numpy/lib/function_base.py

```diff
@@ -175,6 +198,93 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
+            """
+            Rice Estimator
+            Another simple estimator, with no normality assumption.
+            It has better performance for large data, but tends to be too large
+            The number of bins is proportional to the cube root of data size (asymptotically optimal)
+            Depends only on size of the data
+            """
+            return np.ceil(2 * x.size ** (1.0/3))
+
+        def scott(x):
+            """
+            Scott Estimator
+            The binwidth is proportional to the standard deviation of the data and
+            inversely proportional to the cube root of data size (asymptotically optimal)
+
+            """
+            h = 3.5 * x.std() * x.size **(-1.0/3)
+            if h>0:
+                return np.ceil(x.ptp()/h)
+            return 1
+
+        def mad(data, axis=None):
+            """
+            Mean Absolute Deviation - a simple well defined function
+            Simplest implementation, even length case left to default behaviour
+            This is used by the Freedman Diaconis estimator when the IQR does not exist
+            """
+            return np.median(np.absolute(data - np.median(data, axis)), axis)
+
+        def FD(x):
+            """
+            Freedman Diaconis rule using Inter Quartile Range (IQR) for binwidth
+            Considered a variation of the Scott rule with more robustness as the IQR
+            is less affected by outliers than the standard deviation.
+
+            If the IQR is 0, we return the median absolute deviation as defined above, else 1
+            Binwidth is inversely proportional to the cube root of data size (asymptotically optimal)
+            """
+            iqr = np.subtract(*np.percentile(x, [75, 25]))
+
+            if iqr ==0: #unlikely
+                iqr = mad(x) # replace with something useful
+
+            h = (2 * iqr * x.size**(-1.0/3)) # binwidth
+
+            if iqr > 0:
+                return np.ceil(x.ptp()/h)
+
+            return 1 #all else fails
+
+        def default(x):
+            """
+            The FD estimator is usually the most robust method, but it tends to be too small
+            for small X. The Sturges estimator is quite good for small (<1000) datasets and is
+            the default in R.
+            This method gives good off the shelf behaviour.
+            """
+            return max(FD(x), sturges(x))
+
+        optimalityMethods = {'sturges': sturges, 'rice': rice, 'scott': scott,
+                             'fd': FD, 'default': default, 'auto': default}
+
+        for name, estimator in optimalityMethods.items():
+            if name.startswith(bins.lower()):
+                # Undefined behaviour for bins = "s"
+                bins = estimator(a)
+                break
+        else:
+            # Maybe raise a Warning or something? Printing to stderr for now
+            print("Automatic method '{}' not found, reverting to default".format(bins), file=sys.stderr)
```

Comment:
Also, normally you check for valid input first. So:

    if name.lower() not in optimalityMethods.keys():
        raise ValueError('%s not a valid method for `bins`" % name)

    bins = estimator(a)